speaking-portal-project/src/animationFactory/animationProcessor.ts

```diff
@@ -0,0 +1,155 @@
+
+        // TODO: use for testing. After confirming that blinking and other realism is good, delete this
+        console.log("t=%d, frameDur: %d, lastBlink: %d, idling? %s, lastPoseChange: %d",
+                    currentSec, frameDur, lastBlink, lastPoseChange, idle, lastPoseChange)
+
+        // Start building frame's path
+        frameData.concat(`file ../images/${avatar}/${mouthCues[mouthCuesKey].value}`)
+
+        // Blinking
+        blink = willBlink(currentSec, frameDur, lastBlink)
+        if (blink) {
+            lastBlink += 1
+            frameData.concat('_blink')
+        }
+
+        // TODO: determine if avatar will do a body pose or be in idle animation
+
+        // Idling or Posing
+        if(idle){
+            breathPhase = calculateBreathPhase(currentSec, lastPoseChange, breathPhase)
+            frameData.concat(translateBreathFrame(breathPhase))
+        } else {
+            breathPhase = 0
+        }
+
+        // Finish building frame's path
+        frameData.concat(`.png\noutpoint ${frameDur}\n`)
+
+    }
+    return frameData
+}
+
+function willBlink(currentSec: number, frameDur: string, lastBlink: number) {
+    /**
+     * Determines whether or not the avatar will blink based on frame duration and timings
+     *
+     * @remarks
+     * - 1 blink takes ~0.1s, so we only consider frames less than 0.1s long
+     * - Humans typically blink 15-20 times per min, or once every 3-4 seconds
+     * - We want it to be random and not happen exactly once every 3 seconds, so let's say...
+     *      - there's a 40% chance to blink every 4th second
+     *      - there's a 40% chance to blink every 3rd second
+     *      - and don't blink in the same second
+     *
+     * @param currentSec - the current second (integer) of the animation
+     * @param frameDur - the current frame's duration (decimal)
+     * @param lastBlink - the most recent second (integer) in which the avatar blinked
+     *
+     * @returns A boolean
+     *
+    **/
+    let eyes:boolean = false
+    if (parseInt(frameDur) < 0.1 && lastBlink != currentSec) {
+        let n = Math.random()
+        if ((n <= 0.4 && currentSec % 4 == 0) || (n > 0.4 && n <= 0.8 && currentSec % 3 == 0)) {
+            eyes = true
+        }
+    }
+    return eyes
+
+}
+
+function calculateBreathPhase(currentSec: number, lastPoseChange: number, breathPhase: number) {
+    /**
+     * Determines this frame's breath phase value based on time and most recent pose
+     *
+     * @remarks
+     * - A normal respiratory rate in healthy adults is roughly 12 to 20 breaths per minute
+     * - Our avatars' breathing cycle will take 4 seconds, which results in ~15 breaths/min
+     * - Breathing Phases:
+     *      0 = rest
+     *      1 = inhaling or "breathing"
+     *      2 = lungs full
+     *      3 = exhaling or "breathing"
+     *
+     * @returns the new value of breathPhase
+     *
+     **/
```

Comment:
A task we might want to setup for next week is creating some documentation including things like these comments - we could even just move the comments from the code into their own separate documentation document?

Reply:
Yeah, it's a lot of info for a function definition, but I wanted to keep in mind what I was doing while writing the code. We can definitely move stuff around!!